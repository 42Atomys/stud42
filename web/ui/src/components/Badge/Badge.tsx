import classNames from "classnames";
import { Children } from "react";
import { ClassNameProps } from "types/globals";
import BadgeProps, { BadgeChildrenProps, BadgeTextProps } from "./types";

const colorClasses = {
  'purple': 'bg-indigo-500/20 border-indigo-500 text-indigo-700 dark:text-indigo-300 hover:bg-indigo-500/50',
  'blue': 'bg-cyan-500/20 border-cyan-500 text-cyan-700 dark:text-cyan-300 hover:bg-cyan-500/50',
  'green': 'bg-emerald-500/20 border-emerald-500 text-emerald-700 dark:text-emerald-300 hover:bg-emerald-500/50',
  'yellow': 'bg-yellow-500/20 border-yellow-500 text-yellow-700 dark:text-yellow-300 hover:bg-yellow-500/50',
  'red': 'bg-red-500/20 border-red-500 text-red-700 dark:text-red-300 hover:bg-red-500/50',
  'orange': 'bg-orange-500/20 border-orange-500 text-orange-700 dark:text-orange-300 hover:bg-orange-500/50',
  'gray': 'bg-slate-500/20 border-slate-500 text-slate-700 dark:text-slate-300 hover:bg-slate-500/50',
  'white': 'bg-gray-200/20 border-gray-200 text-gray-700 dark:text-gray-300 hover:bg-gray-200/50',
  'black': 'bg-gray-900/20 border-gray-900 text-gray-700 dark:text-gray-300 hover:bg-gray-900/50',
  'transparent': 'bg-transparent border-transparent'
}

export const Badge = ({ ...props }: BadgeProps) => {
  const { color } = props as BadgeTextProps | BadgeChildrenProps
  const { text } = props as BadgeTextProps;
  const { children } = props as BadgeChildrenProps;
  const { className } = props as ClassNameProps

  return <div className={classNames('transition-colors flex w-fit text-gray items-center rounded-full border py-1 px-2 mt-2 text-sm', colorClasses[color], className)}>
    {Children.map(children, c => <>{c}</>) || <span>{text}</span>}
  </div>
}