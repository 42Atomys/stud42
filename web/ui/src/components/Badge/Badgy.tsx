import classNames from 'classnames';
import { PropsWithClassName } from 'types/globals';

/**
 * Badgy is a small badge that can be used to indicate a status or a label
 */
export const Badgy: React.FC<PropsWithClassName<{ text: string }>> = ({
  text,
  className,
}) => (
  <span
    className={classNames(
      'uppercase rounded-full py-0.5 px-2 text-xs font-bold',
      className,
    )}
  >
    {text}
  </span>
);

/**
 * NewBadgy is a small badge that can be used to indicate a status or a label
 * for new features
 */
export const NewBadgy = Badgy.bind(null, {
  text: 'New',
  className: 'bg-indigo-400 dark:bg-indigo-600 text-white',
});

/**
 * BetaBadgy is a small badge that can be used to indicate a status or a label
 * for beta features
 */
export const BetaBadgy = Badgy.bind(null, {
  text: 'Beta',
  className: 'bg-yellow-400 dark:bg-yellow-600 text-white',
});

/**
 * DeprecatedBadgy is a small badge that can be used to indicate a status or a
 * label for deprecated features or features that are going to be removed
 * soon
 */
export const DeprecatedBadgy = Badgy.bind(null, {
  text: 'Deprecated',
  className: 'bg-red-400 dark:bg-red-600 text-white',
});

/**
 * AkaBadgy is a small badge that can be used to indicate a status or a label
 * for anything that are "also known as" something else
 */
export const AkaBadgy = Badgy.bind(null, {
  text: 'Aka',
  className: 'bg-slate-400 dark:bg-slate-700 text-white',
});
