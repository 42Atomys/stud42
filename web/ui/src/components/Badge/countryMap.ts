export const countryNameToEmoji: { [key: string]: string } = {
  Turkey: 'ğŸ‡¹ğŸ‡·',
  Switzerland: 'ğŸ‡¨ğŸ‡­',
  Italy: 'ğŸ‡®ğŸ‡¹',
  Luxembourg: 'ğŸ‡±ğŸ‡º',
  'Czech Republic': 'ğŸ‡¨ğŸ‡¿',
  Armenia: 'ğŸ‡¦ğŸ‡²',
  'Russian Federation': 'ğŸ‡·ğŸ‡º',
  'United Kingdom': 'ğŸ‡¬ğŸ‡§',
  Netherlands: 'ğŸ‡³ğŸ‡±',
  Romania: 'ğŸ‡·ğŸ‡´',
  Brazil: 'ğŸ‡§ğŸ‡·',
  Jordan: 'ğŸ‡¯ğŸ‡´',
  Austria: 'ğŸ‡¦ğŸ‡¹',
  Australia: 'ğŸ‡¦ğŸ‡º',
  Germany: 'ğŸ‡©ğŸ‡ª',
  Canada: 'ğŸ‡¨ğŸ‡¦',
  Finland: 'ğŸ‡«ğŸ‡®',
  Portugal: 'ğŸ‡µğŸ‡¹',
  Ukraine: 'ğŸ‡ºğŸ‡¦',
  'Moldova, Republic of': 'ğŸ‡²ğŸ‡©',
  Spain: 'ğŸ‡ªğŸ‡¸',
  'United Arab Emirates': 'ğŸ‡¦ğŸ‡ª',
  Belgium: 'ğŸ‡§ğŸ‡ª',
  Morocco: 'ğŸ‡²ğŸ‡¦',
  France: 'ğŸ‡«ğŸ‡·',
  'Korea, Republic of': 'ğŸ‡°ğŸ‡·',
  'South Africa': 'ğŸ‡¿ğŸ‡¦',
  Malaysia: 'ğŸ‡²ğŸ‡¾',
  Japan: 'ğŸ‡¯ğŸ‡µ',
  Thailand: 'ğŸ‡¹ğŸ‡­',
  'United States': 'ğŸ‡ºğŸ‡¸',
};
