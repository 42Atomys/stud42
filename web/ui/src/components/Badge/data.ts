import type { FlagDataMap, ThridPartyAccountDataMap } from './types';

/**
 * countryNameToEmoji is a map to convert country name to emoji flag
 * @example countryNameToEmoji['France'] => 'ðŸ‡«ðŸ‡·'
 */
export const countryNameToEmoji: { [key: string]: string } = {
  Turkey: 'ðŸ‡¹ðŸ‡·',
  Switzerland: 'ðŸ‡¨ðŸ‡­',
  Italy: 'ðŸ‡®ðŸ‡¹',
  Luxembourg: 'ðŸ‡±ðŸ‡º',
  'Czech Republic': 'ðŸ‡¨ðŸ‡¿',
  Armenia: 'ðŸ‡¦ðŸ‡²',
  'Russian Federation': 'ðŸ‡·ðŸ‡º',
  'United Kingdom': 'ðŸ‡¬ðŸ‡§',
  Netherlands: 'ðŸ‡³ðŸ‡±',
  Romania: 'ðŸ‡·ðŸ‡´',
  Brazil: 'ðŸ‡§ðŸ‡·',
  Jordan: 'ðŸ‡¯ðŸ‡´',
  Austria: 'ðŸ‡¦ðŸ‡¹',
  Australia: 'ðŸ‡¦ðŸ‡º',
  Germany: 'ðŸ‡©ðŸ‡ª',
  Canada: 'ðŸ‡¨ðŸ‡¦',
  Finland: 'ðŸ‡«ðŸ‡®',
  Portugal: 'ðŸ‡µðŸ‡¹',
  Ukraine: 'ðŸ‡ºðŸ‡¦',
  'Moldova, Republic of': 'ðŸ‡²ðŸ‡©',
  Spain: 'ðŸ‡ªðŸ‡¸',
  'United Arab Emirates': 'ðŸ‡¦ðŸ‡ª',
  Belgium: 'ðŸ‡§ðŸ‡ª',
  Morocco: 'ðŸ‡²ðŸ‡¦',
  France: 'ðŸ‡«ðŸ‡·',
  'Korea, Republic of': 'ðŸ‡°ðŸ‡·',
  'South Africa': 'ðŸ‡¿ðŸ‡¦',
  Malaysia: 'ðŸ‡²ðŸ‡¾',
  Japan: 'ðŸ‡¯ðŸ‡µ',
  Thailand: 'ðŸ‡¹ðŸ‡­',
  'United States': 'ðŸ‡ºðŸ‡¸',
};

/**
 * flagData is a map how contains data about UserFlag and how to display it
 * @example flagData[UserFlag.STAFF] => { ...  }
 */
export const flagData: FlagDataMap = {
  STAFF: {
    name: 'Staff',
    description: 'S42 staff member',
    className: 'fa-duotone fa-hat-wizard text-indigo-500',
  },
  BETA: {
    name: 'Beta Tester',
    description: 'Present during the beta phase',
    className: 'fa-duotone fa-flask-potion text-orange-500',
  },
  COLLABORATOR: {
    name: 'Collaborator',
    description: 'Collaborator on Github',
    className: 'fa-duotone fa-code-compare text-cyan-500',
    link: () => new URL('https://github.com/42Atomys/stud42/collaborators'),
  },
  CONTRIBUTOR: {
    name: 'Contributor',
    description: 'Contributed on Github',
    className: 'fa-duotone fa-code-pull-request text-green-500',
    link: () => new URL('https://github.com/42Atomys/stud42'),
  },
  DISCORD: {
    name: 'Discord',
    description: 'Member of the Discord server',
    className: 'fa-brands fa-discord text-slate-500',
    link: () => new URL('https://discord.gg/5f864c6hyj'),
  },
  SPONSOR: {
    name: 'Sponsor',
    description: 'Sponsor of the project',
    className: 'fa-duotone fa-user-astronaut text-pink-500',
    link: () => new URL('https://github.com/sponsors/42Atomys'),
  },
};

/**
 * thridPartyData is a map how contains data about ThridPartyAccount
 * (also called `Account` on API) and how to display it.
 * @example thridPartyData[AccountProvider.DISCORD] => { ...  }
 */
export const thridPartyData: ThridPartyAccountDataMap = {
  DUO: {
    name: 'Intra 42',
    className: 'fa-kit fa-duoquadra',
    link: ({ username }) =>
      new URL(`https://profile.intra.42.fr/users/${username}`),
  },
  SLACK: {
    name: 'Slack',
    className: 'fab fa-slack',
    link: ({ username }) =>
      new URL(`https://42born2code.slack.com/messages/@${username}`),
  },
  DISCORD: {
    name: 'Discord',
    className: 'fab fa-discord',
    link: ({ username }) => new URL(`https://discord.com/users/${username}`),
  },
  GITHUB: {
    name: 'Github',
    className: 'fab fa-github',
    link: ({ username }) => new URL(`https://github.com/${username}`),
  },
  GITLAB: {
    name: 'Gitlab',
    className: 'fab fa-gitlab',
    link: ({ username }) => new URL(`https://gitlab.com/${username}`),
  },
  LINKEDIN: {
    name: 'LinkedIn',
    className: 'fab fa-linkedin',
    link: ({ providerAccountId }) =>
      new URL(`https://www.linkedin.com/in/${providerAccountId}`),
  },
  TWITTER: {
    name: 'Twitter',
    className: 'fab fa-twitter',
    link: ({ username }) => new URL(`https://twitter.com/${username}`),
  },
  INSTAGRAM: {
    name: 'Instagram',
    className: 'fab fa-instagram',
    link: ({ username }) => new URL(`https://www.instagram.com/${username}`),
  },
  REDDIT: {
    name: 'Reddit',
    className: 'fab fa-reddit',
    link: ({ username }) => new URL(`https://www.reddit.com/user/${username}`),
  },
  SPOTIFY: {
    name: 'Spotify',
    className: 'fab fa-spotify',
    link: ({ providerAccountId }) =>
      new URL(`https://open.spotify.com/user/${providerAccountId}`),
  },
  TWITCH: {
    name: 'Twitch',
    className: 'fab fa-twitch',
    link: ({ username }) => new URL(`https://www.twitch.tv/${username}`),
  },
};
