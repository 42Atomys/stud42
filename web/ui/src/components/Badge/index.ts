export { Badge } from './Badge';
export { AkaBadgy, BetaBadgy, DeprecatedBadgy, NewBadgy } from './Badgy';
export { FlagBadge } from './FlagBadge';
export { LocationBadge } from './LocationBadge';
export { ThridPartyBadge } from './ThridPartyBadge';
export { flagData, thridPartyData } from './data';
export type { BadgeColor, BadgeProps } from './types';
export { thirdPartySorted } from './utils';
