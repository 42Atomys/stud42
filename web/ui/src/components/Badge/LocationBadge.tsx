import Emoji from '@components/Emoji';
import { Location } from '@graphql.d';
import classNames from 'classnames';
import { NestedPartial } from 'types/utils';
import { Badge } from './Badge';
import { countryNameToEmoji } from './countryMap';

export const LocationBadge = ({
  location,
}: {
  location: NestedPartial<Location> | null | undefined;
}) => {
  const isConnected = location?.identifier ? true : false;

  return (
    <Badge color={isConnected ? 'green' : 'gray'}>
      <span
        className={classNames(
          'inline-flex rounded-full w-2 h-2',
          isConnected ? 'bg-emerald-500' : 'bg-slate-500'
        )}
      ></span>
      <span className="flex flex-row justify-center items-center text-sm mx-1">
        {isConnected ? location?.identifier : 'Offline'}
      </span>
      {isConnected && (
        <Emoji
          emoji={countryNameToEmoji[location?.campus?.country || '']}
          size={14}
          title={location?.campus?.name}
          className={classNames('mx-1', isConnected ? 'visible' : 'hidden')}
        />
      )}
    </Badge>
  );
};
