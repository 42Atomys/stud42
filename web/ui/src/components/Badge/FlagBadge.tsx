import { ConditionalWrapper } from '@components/ConditionalWrapper';
import { Tooltip } from '@components/Tooltip';
import { UserFlag } from '@graphql.d';
import classNames from 'classnames';
import Link from 'next/link';
import { flagData } from './data';

/**
 * The FlagBadge component is a UI component that displays an icon representing
 * a particular flag along with a tooltip containing more information about that
 * flag.
 *
 * It takes in a flag props, which specifies the type of flag to display.
 * The component uses a data object: `flagData`, to determine how to display it.
 * If the flagData object also contains a link property for the flag,
 * the icon will be wrapped in a Link component that opens the link in a
 * new tab when clicked.
 */
export const FlagBadge: React.FC<{ flag: UserFlag }> = ({ flag }) => {
  const data = flagData[flag];

  return (
    <Tooltip
      text={data.name}
      subText={data.description}
      color="black"
      direction="bottom"
      size="sm"
      allowInteractions={false}
      showArrow
      className="![--tooltip-mb:1rem]"
    >
      <ConditionalWrapper
        condition={data.link !== undefined}
        trueWrapper={(children) => (
          <Link
            role="link"
            href={data.link?.(null) as URL}
            className="flex"
            target="_blank"
          >
            {children}
          </Link>
        )}
      >
        <i
          data-testid={`flag-${flag}`}
          className={classNames('fa-fw cursor-pointer p-1', data.className)}
        />
      </ConditionalWrapper>
    </Tooltip>
  );
};
