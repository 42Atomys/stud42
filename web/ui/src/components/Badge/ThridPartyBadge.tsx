import { ConditionalWrapper } from '@components/ConditionalWrapper';
import { Tooltip } from '@components/Tooltip';
import classNames from 'classnames';
import Link from 'next/link';
import { thridPartyData } from './data';
import { ThridPartyBadgeProps } from './types';

/**
 * ThridPartyUnkwownError is thrown when the provider specified in the props is
 * not supported by the thridPartyData object.
 *
 * When this error is thrown, check the thridPartyData object to see if the
 * provider is supported. If it is, make sure that the provider is spelled
 * correctly in the thridPartyData object. If it is not, add the provider to
 * the thridPartyData object.
 */
export const ThridPartyUnkwownError = new Error('Unknown thrid party provider');

/**
 * This is a UI component called ThridPartyBadge that displays an icon
 * representing a third-party account provider, such as Twitter or GitHub, along
 * with a tooltip containing the name of the provider.
 *
 * The component uses a data object called thridPartyData to determine how to
 * display the badge.
 * If the provider specified in the props is not supported by the thridPartyData
 * object, the component throws the `ThridPartyUnkwownError` error.
 *
 * If a link property exists for the provider in the thridPartyData object,
 * the icon will be wrapped in a Link component that opens the link in a new tab
 * when clicked and redirects the user to the user's profile on that provider's
 * website.
 */
export const ThridPartyBadge: React.FC<ThridPartyBadgeProps> = ({
  provider,
  username,
  providerAccountId,
}) => {
  const data = thridPartyData[provider];
  if (!data) throw ThridPartyUnkwownError;

  return (
    <Tooltip
      text={data.name}
      direction="bottom"
      size="sm"
      color="black"
      allowInteractions={false}
    >
      <ConditionalWrapper
        condition={!!data.link}
        trueWrapper={(children) => (
          <Link
            role="link"
            className="hover:bg-slate-950 rounded-lg p-2"
            key={`user-profile-acount-tooltip-${providerAccountId}`}
            href={data.link?.(username) as URL}
          >
            {children}
          </Link>
        )}
      >
        <i
          data-testid={`thrid-party-${provider}`}
          className={classNames('fa-fw', data.className)}
        />
      </ConditionalWrapper>
    </Tooltip>
  );
};
