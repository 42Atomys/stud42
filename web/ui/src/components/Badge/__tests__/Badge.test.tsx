import { Badge, BadgeColor } from '@components/Badge';
import { render } from '@testing-library/react';

describe('snapshots', () => {
  it('renders Badge unchanged', () => {
    const { container } = render(<Badge color="black" text="Testing" />);
    expect(container).toMatchSnapshot();
  });

  it('renders Badge with children unchanged', () => {
    const { container } = render(
      <Badge color="black">
        <i>Testing</i>
      </Badge>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Badge with className unchanged', () => {
    const { container } = render(
      <Badge color="black" text="Testing" className="test" />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('color data', () => {
  [
    { color: 'purple', class: 'indigo', opacity: '500' },
    { color: 'fuchsia', class: 'fuchsia', opacity: '500' },
    { color: 'blue', class: 'cyan', opacity: '500' },
    { color: 'green', class: 'emerald', opacity: '500' },
    { color: 'yellow', class: 'yellow', opacity: '500' },
    { color: 'red', class: 'red', opacity: '500' },
    { color: 'orange', class: 'orange', opacity: '500' },
    { color: 'gray', class: 'slate', opacity: 500 },
    { color: 'white', class: 'gray', opacity: 200 },
    { color: 'black', class: 'gray', opacity: 900 },
  ].forEach(({ color, class: className, opacity }) => {
    it(`renders color data with ${color}`, () => {
      const { container } = render(
        <Badge color={color as BadgeColor} text="Testing" />,
      );
      expect(container.firstChild).toHaveClass(`bg-${className}-${opacity}/20`);
      expect(container.firstChild).toHaveClass(
        `border-${className}-${opacity}`,
      );
      expect(container.firstChild).toHaveClass(`text-${className}-700`);
      expect(container.firstChild).toHaveClass(`dark:text-${className}-300`);
      expect(container.firstChild).toHaveClass(
        `hover:bg-${className}-${opacity}/50`,
      );
    });
  });

  it('renders color data with transparent', () => {
    const { container } = render(<Badge color="transparent" text="Testing" />);
    expect(container.firstChild).toHaveClass('bg-transparent');
    expect(container.firstChild).toHaveClass('border-transparent');
  });
});

describe('children and text', () => {
  it('renders color data with text', () => {
    const { container } = render(<Badge color="transparent" text="Testing" />);

    expect(container.firstChild).toBeInTheDocument();
    expect(container.firstChild).toContainHTML('Testing');
  });

  it('renders color data with children', () => {
    const { container } = render(
      <Badge color="transparent">
        <i>Testing</i>
      </Badge>,
    );

    expect(container.firstChild).toBeInTheDocument();
    expect(container.firstChild).toContainHTML('<i>Testing</i>');
  });

  it('render color data with multiple children', () => {
    const { container } = render(
      <Badge color="transparent">
        <i>Testing 1</i>
        <b>Testing 2</b>
      </Badge>,
    );

    expect(container.firstChild).toBeInTheDocument();
    expect(container.firstChild).toContainHTML(
      '<i>Testing 1</i><b>Testing 2</b>',
    );
  });
});
