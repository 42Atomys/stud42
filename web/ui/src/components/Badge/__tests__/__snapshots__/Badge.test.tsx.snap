// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders Badge unchanged 1`] = `
<div>
  <div
    class="transition-colors flex w-fit text-gray items-center rounded-full border py-1 px-2 mt-2 text-sm bg-gray-900/20 border-gray-900 text-gray-700 dark:text-gray-300 hover:bg-gray-900/50"
    data-testid="badge"
  >
    <span>
      Testing
    </span>
  </div>
</div>
`;

exports[`snapshots renders Badge with children unchanged 1`] = `
<div>
  <div
    class="transition-colors flex w-fit text-gray items-center rounded-full border py-1 px-2 mt-2 text-sm bg-gray-900/20 border-gray-900 text-gray-700 dark:text-gray-300 hover:bg-gray-900/50"
    data-testid="badge"
  >
    <i>
      Testing
    </i>
  </div>
</div>
`;

exports[`snapshots renders Badge with className unchanged 1`] = `
<div>
  <div
    class="transition-colors flex w-fit text-gray items-center rounded-full border py-1 px-2 mt-2 text-sm bg-gray-900/20 border-gray-900 text-gray-700 dark:text-gray-300 hover:bg-gray-900/50 test"
    data-testid="badge"
  >
    <span>
      Testing
    </span>
  </div>
</div>
`;
